File: leetcode/questions/399/python/399.py
from typing import List

from collections import defaultdict, deque

class Solution:
    def calcEquation(self, equations: List[List[str]], values: List[float], queries: List[List[str]]) -> List[float]:
        def calculate(graph, st, end):
            marked = set()
            queue = deque([(st, 1)])

            while queue:
                cur, val = queue.popleft()

                if cur == end:
                    return val

                if cur in marked:
                    continue

                marked.add(cur)

                for nxt, nxt_val in graph[cur]:
                    queue.append((nxt, nxt_val * val))

            return -1
        
        graph = defaultdict(list)

        for idx, (st, end) in enumerate(equations):
            graph[st].append((end, values[idx]))
            graph[end].append((st, 1/values[idx]))

        ans = []

        for st, end in queries:
            if st not in graph or end not in graph:
                ans.append(-1)
            else:
                ans.append(calculate(graph, st, end))

        return ans

File: leetcode/questions/399/java/399.java
import java.util.*;

class Solution {
    class Pair {
        public String node;
        public double val;

        public Pair(String node, double val) {
            this.node = node;
            this.val = val;
        }

        @Override
        public String toString() {
            return String.format("%s %.2f", this.node, this.val);
        }
    }

    private double calculate(String dividend, String divisor, HashMap<String, HashMap<String, Double>> graph) {
        Stack<Pair> stack = new Stack<>();
        HashSet<String> marked = new HashSet<>();
        stack.push(new Pair(dividend, 1.0));
        

        while (!stack.isEmpty()) {
            Pair cur = stack.pop();

            if (divisor.equals(cur.node)) {
                return cur.val;
            }

            if (marked.contains(cur.node)) {
                continue;
            }

            marked.add(cur.node);
            graph.get(cur.node).forEach((dest, val) -> {
                stack.push(new Pair(dest, cur.val * val));
            });

        }

        return -1.0;
    }

    public double[] calcEquation(List<List<String>> equations, double[] values, List<List<String>> queries) {
        HashMap<String, HashMap<String, Double>> graph = new HashMap<String, HashMap<String, Double>>();

        for (int i = 0; i < equations.size(); ++i) {
            String dividend = equations.get(i).get(0);
            String divisor = equations.get(i).get(1);
            double quocient = values[i];

            if (!graph.containsKey(dividend)) {
                graph.put(dividend, new HashMap<String, Double>());
            }
            if (!graph.containsKey(divisor)) {
                graph.put(divisor, new HashMap<String, Double>());
            }

            graph.get(dividend).put(divisor, quocient);
            graph.get(divisor).put(dividend, 1/quocient);
        }

        double[] ans = new double[queries.size()];

        for (int i = 0; i < queries.size(); ++i) {
            String dividend = queries.get(i).get(0);
            String divisor = queries.get(i).get(1);

            if(!graph.containsKey(dividend) || !graph.containsKey(divisor)) {
                ans[i] = -1;
            }

            else {
                ans[i] = calculate(dividend, divisor, graph);
            }
        }

        return ans;
    }
}